File: 16_3Sum Closest.java
// Two Pointers
// Time complexity: O(n^2)
// Space complexity: 
//      O(1) or O(n) extra space depending on the sorting algorithm.
//      O(m) space for the output list.

class Solution {
    public int threeSumClosest(int[] nums, int target) {
        Arrays.sort(nums);
        int res = nums[0] + nums[1] + nums[2];
        for (int i = 0; i < nums.length - 2; i++) {
            // edge case 1: cur + min 2 > target -> the following will definitely even larger and further
            int opt = nums[i] + nums[i+1] + nums[i+2];
            if (opt > target) {
                if (Math.abs(target - opt) < Math.abs(target - res)) res = opt;
                break;
            }
            // edge case 2: cur + max 2 < target -> the following will definitely even smaller and further
            opt = nums[i] + nums[nums.length - 2] + nums[nums.length - 1];
            if (opt < target) {
                if (Math.abs(target - opt) < Math.abs(target - res)) res = opt;
                continue;
            }
            // edge case 3: avoid repeated calculation
            if (i > 0 && nums[i] == nums[i-1]) continue;

            int left = i + 1, right = nums.length - 1;
            while (left < right) {
                int tmp = nums[i] + nums[left] + nums[right];
                if (tmp == target) return tmp;
                if (Math.abs(target - tmp) < Math.abs(target - res)) res = tmp;

                // shift pointers
                if (tmp < target) left++;
                else right--;
            }
        }
        return res;
    }
}

File: 16_3Sum Closest.py
class Solution:
    def threeSumClosest(self, nums: List[int], target: int) -> int:
        # 1. sort
        nums.sort()

        # 2. 2Sum
        res = inf

        for i in range(len(nums) - 2):
            # Trim branch
            # case: min_sum > target => more diff later => terminate
            min_sum = nums[i] + nums[i+1] + nums[i+2]
            if min_sum > target:
                if abs(target - min_sum) < abs(target - res): res = min_sum
                break
            # case: max_sum < target => skip & increase i ASAP
            max_sum = nums[i] + nums[-1] + nums[-2]
            if max_sum < target:
                if abs(target - max_sum) < abs(target - res): res = max_sum
                continue

            l, r = i + 1, len(nums) - 1
            while l < r:
                tsum = nums[i] + nums[l] + nums[r]
                if abs(target - tsum) < abs(target - res): res = tsum

                if tsum > target: r -= 1
                elif tsum < target: l += 1
                else: break
        return res